### PR TITLE
[SYCL] Replace CUDA & HIP features.hpp headers with macros

### PR DESCRIPTION
**Pull request by kbenzie**
Split out from https://github.com/intel/llvm/pull/14145 this patch replaces the generated `features.hpp` header for the CUDA and HIP plugins, which only defined single feature test macro, with `#cmakedefine01` definitions controlled by the build system.

### CODE REVIEW

File: sycl/CMakeLists.txt
Comment (AlexeySachkov) [inline sycl/CMakeLists.txt:160]: Would it be better to add `adapter` here right away? Or maybe something like `backend`? Because it may be a bit confusing what does it mean "sycl build cuda".

But that's just a though, I will leave a final word to the code owners.

File: sycl/test/CMakeLists.txt
Comment (AlexeySachkov) [inline sycl/test/CMakeLists.txt:97]: Just an observation: I don't think that `sycl/test` should be repeated for different triples. If there are some integration tests which check some intermediates produced by the compiler and they depend on the triple - they should set it explicitly.

Otherwise, tests which are supposed to be device-agnostic should not be run for each triple individually.

I created #14671 to track this.

File: sycl/CMakeLists.txt
Comment (kbenzie) [reply sycl/CMakeLists.txt:160]: I think using `backend` terminoligy in the SYCL build would make most sense, personally.

**Review by AlexeySachkov — APPROVED** [review]

**Comment by kbenzie** [comment]
@uditagarwal97 anything needing changed or is this in a good state for apporoval?

**Review by uditagarwal97 — APPROVED** [review]
LGTM

File: sycl/CMakeLists.txt
Comment (keryell) [inline sycl/CMakeLists.txt:171]: `SYCL_BUILD_BACKEND_LEVEL_ZERO`

**Pull request merged**
